### ISSUE TITLE
Future of nomacs and call for help/volunteers

### ISSUE THREAD

**Issue by novomesk** [nomacs/nomacs#987]
Hello,

Future of nomacs depends on people who wish nomacs to exist in the future.

nomacs has lot of features, it is not a small project. There is lot of work to do, nomacs needs a team of maintainers. Current rate of development is better than nothing but insufficient for a long term perspective.

The work is needed mostly in the following areas:

- Maintenance (to make sure nomacs builds when its dependencies upgrade, refactor obsolete code)
- Review bug reports and provide solutions/fixes.
- Review Pull requests from one-time contributors
- Implement requested features, improvements
- Port for macOS. 
- Testing

I think it will be necessary to fork the project so that we have full control about everything. That is technically not a problem to do, the problem to find people for the team.

The team needs both **developers** and **testers**.

**1) Developers**

I understand that everyone is busy and free time is precious. Infrequent commits are fine but there is need for long-term help, perseverance and to take responsibility for certain areas of code. Lack of previous experience is not a problem, the project is a great opportunity to learn.

**2) Testers**

The testing is also very important and needed. For example when some developer implements big changes, it is very useful to verify if there are no regressions.

We would need:

- Windows user(s) who is willing to try development (experimental, unreleased) builds and to provide feedback when needed.
- Advanced Linux users who are capable to build/compile nomacs from git, install and test and to provide feedback when needed.

If you are willing to keep helping the nomacs project and to be a member of the potential future team, please notify me under this issue. Tell me what you will like to do (as a developer or tester) and what is your motivation (why you wish to help the nomacs project).

If we don't form a team and don't maintain nomacs properly, I guess that nomacs would be gradually removed from many repositories and users will move to other image viewers as a consequence.

**Comment by Mazoch** [comment]
I may be littering the thread for which I apologize, but I would just like to say that I have tested several different photo viewers and this one seems the best to me because it works. Every format launches without a problem. Thank you very much for the work you put into this project.

**Comment by CSA-IT-Team** [comment]
I would be happy to be a Windows tester however at present I would need the builds to be created so I can install them. I can offer testing in a desktop and RDP environment (Windows Server 2022).

**Comment by novomesk** [comment]
@CSA-IT-Team Thank you.

The builds for Windows are created automatically almost after each commit on AppVeyor. Also when someone make a pull request with new features, there is a build on AppVeyor. The builds are [available](https://ci.appveyor.com/project/diemmarkus/nomacs/builds/48017903/artifacts) publicly 1 month. I'll notify you when I need to test some changes.

**Comment by Galaxoid-333** [comment]
The main advantage of Nomacs is that it is fast. But there are also shortcomings that make work difficult. In fact, I know of two similar image viewing applications: Nomacs and ImageGlass. Other applications are not suitable at all.

p.s. It would be good to think about changing the name to something more intelligible.

**Comment by Serl0p** [comment]
I've been using Nomacs for the last year and really love it. I have a Windows 11 laptop, and would be happy to help with testing! If I knew how to code I'd love to help develop, but I have no experience 😅

**Comment by leejuyuu** [comment]
@novomesk Hello, thanks for keeping this project alive! Nomacs has been my favorite image viewer, and I would love to help with its maintenance. Here's where I think I can help:

- Start triaging the issues. Add tags to help categorize them better, and close duplicates or non-reproducible ones. I would probably need privileges to modify the issues though.
- I know C++ though I am not very fluent. I can help fix some simpler bugs (slowly).
- I work on web development in my day job. I'm probably more backend-oriented but could do basic HTML and CSS, so might be able to help resurrect the Nomacs website.

**Comment by novomesk** [comment]
@leejuyuu Your willingness to help is welcomed.

Regarding the work on website, please contact @mmahmoudian who has access to the domain/website.

**Comment by mmahmoudian** [comment]
@leejuyuu @novomesk the issue I have been facing at the moment is making the Github Action to work:

https://github.com/nomacs/nomacs.github.io/blob/main/.github/workflows/build.yml

**Comment by mmahmoudian** [comment]
@leejuyuu I managed to fix the CI/CD and it now works. A minimal website for the test is up. I think from this point on wards it would be straightforward to contribute to the website like any other Github repo.

**Comment by leejuyuu** [comment]
@novomesk Is it possible to add more tags? I would like to have "docs" and "translation" as well so that it can be easier for contributors to find.

**Comment by Galaxoid-333** [comment]
> @leejuyuu I managed to fix the CI/CD and it now works. A minimal website for the test is up. I think from this point on wards it would be straightforward to contribute to the website like any other Github repo.

Now we need more people to know about this new repository for Nomacs.

**Comment by novomesk** [comment]
@leejuyuu you can add the tags you need.

**Comment by leejuyuu** [comment]
@novomesk Thanks!

**Comment by leejuyuu** [comment]
Another question, do you think discussions about documentation belong in this repo or the nomacs.github.io repo? Managing them in one place might make them easier to find, but now that the codes are separated.

**Comment by mmahmoudian** [comment]
@leejuyuu In my opinion we should consider two points:
1. They were not previously separated either, so same practice as before with the exception that there is a better link between the website and the code repo
2. We should consider the audience. Which layout/platform would they feel more comfortable with when reading documentation. Github or website.

imho the very technical documentations can live in the main repo (and perhaps mirrored on the website), and the general documentation live on the website. Then the only issue left would be how to define what is technical and what is general.

**Comment by leejuyuu** [comment]
I probably should phrase it like this:
Should the issues about the user-facing documentation live in the main repo or the website repo? 

It seems more likely that users report things in the main repo and we probably should keep it that way. Then the problem is whether to move the issues. I'm currently thinking that it might not be necessary, as long as they can be referenced from another repo.

**Comment by leejuyuu** [comment]
> > I probably should phrase it like this: Should the issues about the user-facing documentation live in the main repo or the website repo?
> > 
> > It seems more likely that users report things in the main repo and we probably should keep it that way. Then the problem is whether to move the issues. I'm currently thinking that it might not be necessary, as long as they can be referenced from another repo.
> 
> Also if your talking about moving the issues from the main repo you could do it like my favorite adblocker UblockOrgin and look at their repo for an example on how they did it @leejuyuu 

Thanks for the suggestion, but my question is about whether to move, not how to move.

**Comment by novomesk** [comment]
@Coldtrain320 

> My suggestion is updating the latest bit of the code to qt6 since a lot of people use qt6 now a days. Sorry if you all hate my suggestion.

The suggestion to port code to Qt6 is good. We know that KDE is moving to Qt6 too and Qt5 is becoming obsolete.

We just need people who would turn that suggestion into reality - developers and testers.

If you are on Windows, you can help us by testing latest development build https://ci.appveyor.com/project/diemmarkus/nomacs/builds/49142263/artifacts and to check if there are not regressions compared to previous releases.

Preparations for Qt6 already started. However recent changes has to be tested/verified by experienced users. Sometimes it happens that new changes create new bugs.

**Comment by ppenguin** [comment]
> The suggestion to port code to Qt6 is good. We know that KDE is moving to Qt6 too and Qt5 is becoming obsolete.
> We just need people who would turn that suggestion into reality - developers and testers.
 
> Preparations for Qt6 already started. However recent changes has to be tested/verified by experienced users. 

I'll make a nix package `nomacs-qt6`, which will enable nix users to easily run e.g. `master` builds in production to make real-life testing easier.

The prerequisite for this is that a qt6-based development build from a current branch can be built in the "usual" way and works generally.

**Comment by novomesk** [comment]
> Anyway to add .dds format viewing on the windows application ?

The question is not related with subject of this thread.

If you want, build this [plug-in](https://github.com/qt/qtimageformats/tree/5.15/src/plugins/imageformats/dds) and add your qdds.dll to enable viewing of the DDS files.

Here is the info why the plug-in is not built by default: https://github.com/qt/qtimageformats/commit/06ee5a2abc560a1041d2c9f80eaa42f5de80a4f9

**Comment by raxleys** [comment]
@novomesk I have ~3 years of professional C++ experience, but I've never worked with QT or on a cross-platform application (only developed for Linux).

Nomacs is one of my favorite image viewers, and it'd be great to see it back in my distro's repos one day. I'm interested in contributing as a developer (feature implementation/bug fixes), but am wondering what kind of commitment would be expected - I've never contributed to open source before.

**Comment by novomesk** [comment]
@raxleys You yourself decide how many hours per month you want to dedicate to the project. There are many things to do, there are unfixed bugs, feature requests, etc. Choose anythings and discuss in the Matrix chat room what you would like to do next.

**Comment by raxleys** [comment]
@novomesk thanks, I'll start by looking through the issues then

**Comment by scrubbbbs** [comment]
I am new to the code (about one week), I think a fork is a good idea if the core devs are hesitant or simply unavailable to promote anyone to merge PRs or be an active committer.

I like the software, have been using it forever but never tried fixing anything until recently. I am mostly interested in improving zip file support and have been through most of the core code looking at how to solve it, and have already hacked together a branch with a lot of improvements. But I'm not sure if I can dice it up into small enough PRs to get it merged or will have the patience for a drawn out process.

I am not a fan of the code at the moment, it seems too verbose and pedantic, a bit saturated with redundant sanity checks and yet virtually no assertions. There is a lot of dead and trivial code too - which is a matter of taste I know but for a project of this caliber it seems like too much. While fixing a segfault today I also found out the way threads are being used is not great either which can wreak all sorts of havok.

With all that said I am leaning towards feature freezing the project until a there is a major bug-hunt effort, refactoring effort, all new bugs are squashed and a lot of the cruft can be deleted. I don't think it is in a state to pile on new features or port to Qt6 for example. To survive I think it needs a really stable and glitch-free release right now.

**Comment by paulocoghi** [comment]
Would be nice to add the site URL nomacs.org to the repository description here on GitHub

**Comment by leejuyuu** [comment]
@paulocoghi  Thanks for the suggestion! It's already in the readme, but it would also be nice to have it on the sidebar.

@novomesk I do not have access to the repo settings, would you mind editing the About section of the repo?

![image](https://github.com/user-attachments/assets/3d218fc7-70bc-4999-80db-64238ff59818)

**Comment by mmahmoudian** [comment]
@paulocoghi 

> Would be nice to add the site URL nomacs.org to the repository description here on GitHub

Thanks for the suggestion. It is now done.

**Comment by c0xc** [comment]
Hi there, I've made two major changes two years ago because I wanted to improve Nomacs as I use it myself. My initial motivation was to make further improvements ([1](https://github.com/nomacs/nomacs/issues/804)) but when I saw certain parts of the code plus due to time constraints, I changed my mind.

However, I'll try to be there and help if someone has a question or an issue with the specific parts that I have changed. I might be able to have a tip or two if someone is having issues building Nomacs on Linux too, as I did all that manually, but that's about it for now.

My changes were (+ (lost) motivation):

* PR #837 - I fixed the bug where Nomacs silently modified a file on disk without asking the user after rotating it. This was dangerous, hit my several times plus I even had a case when someone had lost some data that was linked to an image in a datatabase because that image file was then modified (when I found all the other files in the dir were 10 years old except this one, it was obvious). 
This whole process was why I lost my motivation for further changes, when I saw how many wrappers are used, some data is stored in one some in the other, which lead me to change way more than I wanted to change. A bug when changing notes (sometimes cleared), I had to extend the undo history to make undo possible at all, then some random crash because one of the many wrappers would be reused and modified in parallel (when browsing) instead of atomically being loaded once and discarded... Especially that last part was why I decided not to make further contributions as the code, at least these sections were too error-prone or require lots of refactoring.
* Issue #850 - There was a new yet incomplete and very buggy crop feature which essentially made it almost impossible to crop certain images. I was hesitant reverting someone else's work but it was obviously worse than the old one and after reading many comments complaining about it, I decided to revert it and restore the old crop.

**Comment by paulocoghi** [comment]
> I'll try to be there and help if someone has a question or an issue with the specific parts that I have changed

@c0xc thanks a lot for coming back!

**Comment by ppenguin** [comment]
> I'll make a nix package `nomacs-qt6`, which will enable nix users to easily run e.g. `master` builds in production to make real-life testing easier.

Great job on making the build Qt* functional: it was now a breeze to add a Qt6 version: https://github.com/NixOS/nixpkgs/pull/359160

**Comment by scrubbbbs** [comment]
Nomacs 3.19 is in Debian sid/unstable as of 2024/12/8, thanks to everyone who helped!

**Comment by jo-chemla** [comment]
Great to hear! Can this 3.19 version be published or released as github release for windows as well - if it makes any sense?

**Comment by scrubbbbs** [comment]
> Great to hear! Can this 3.19 version

There will probably be another beta release on github soon (3.19.2) which would include the version debian pulled from. There are a couple off PRs we should take care of first.

**Comment by jo-chemla** [comment]
Thanks for the comment, good luck with the ongoing work reviving nomacs!

**Comment by j-jasz** [comment]
I just switched from Viewnior to nomacs on Debian 13 and it's really nice and so much faster. Congrats!

I can occasionally compile nomacs and test some features, but it will depend largely on my work schedule.
For day to day use I will stick to the version in Debian repositories.

**Comment by scrubbbbs** [comment]
> I can occasionally compile nomacs and test some features, but it will depend largely on my work schedule.

We have Linux app images in CI now. You can find them in [Actions/Linux Workflow](https://github.com/nomacs/nomacs/actions/workflows/linux.yml). From there choose a workflow run and download one of the zip files. The 24.04 quazip build includes all features and image plugins.
